File: webhooks/nodejs/samples/room.deleted.js
// Room Deletion - Notify Participants, Admin, and Save Access Log

const express = require("express");
const crypto = require("crypto");
const axios = require("axios");
const fs = require("fs");
const path = require("path");
const FormData = require("form-data");
require("dotenv").config();

const app = express();
app.use(express.json({ verify: rawBodySaver }));

function rawBodySaver(req, res, buf) {
  req.rawBody = buf.toString();
}

function getSecretHash(secret, payload) {
  return (
    "sha256=" +
    crypto.createHmac("sha256", secret).update(payload).digest("hex")
  );
}

const SENSITIVE_ROOM_KEYWORDS = ["hr", "finance", "legal", "clients", "external"];

app.post("/webhook", async (req, res) => {
  const signature = req.headers["x-docspace-signature-256"];
  const computedSignature = getSecretHash(
    process.env.WEBHOOK_SECRET_KEY,
    req.rawBody
  );
  if (signature !== computedSignature) return res.sendStatus(401);

  const { trigger } = req.body.event;
  if (trigger !== "room.deleted") return res.sendStatus(200);

  const { title: roomTitle, rootId: roomId, createBy } = req.body.payload;
  const lowerTitle = roomTitle.toLowerCase();
  const timestamp = new Date().toISOString();

  console.log(`Room deleted: "${roomTitle}" (${roomId})`);

  try {
    // Step 1: Get room participants and access types
    const shareRes = await axios.get(
      `${process.env.PORTAL_URL}/api/2.0/files/rooms/${roomId}/share`,
      {
        headers: { Authorization: `Bearer ${process.env.API_KEY}` }
      }
    );

    const users = shareRes.data.access || [];

    // Step 2: Notify all participants
    for (const user of users) {
      const email = user.sharedTo?.email;
      if (!email) continue;

      await axios.post(
        process.env.NOTIFY_WEBHOOK_URL,
        {
          to: email,
          subject: "Room deleted",
          message: `The room "${roomTitle}" has been deleted. If this is unexpected, please contact your administrator.`
        },
        { headers: { "Content-Type": "application/json" } }
      );

      console.log(`User notified: ${email}`);
    }

    // Step 3: Save deletion access log
    const accessLog = users.map(user => ({
      email: user.sharedTo?.email || "[no email]",
      accessLevel: user.access || "[unknown]"
    }));

    const logData = {
      roomId,
      roomTitle,
      deletedAt: timestamp,
      accessBeforeDeletion: accessLog
    };

    const logFileName = `deleted_room_${roomId}_${timestamp.replace(/[:.]/g, "-")}.json`;
    const logPath = path.join(__dirname, logFileName);

    fs.writeFileSync(logPath, JSON.stringify(logData, null, 2));

    const form = new FormData();
    form.append("file", fs.createReadStream(logPath));

    const uploadRes = await axios.post(
      `${process.env.PORTAL_URL}/api/2.0/files/folder/${process.env.DELETION_LOG_FOLDER_ID}/upload`,
      form,
      {
        headers: {
          Authorization: `Bearer ${process.env.API_KEY}`,
          ...form.getHeaders()
        }
      }
    );

    const uploaded = uploadRes.data?.response?.[0];
    const logUrl = `${process.env.PORTAL_URL}/file/${uploaded.id}`;

    fs.unlinkSync(logPath);

    console.log("Access log uploaded");

    // Step 4: If sensitive room - notify admin
    const isSensitive = SENSITIVE_ROOM_KEYWORDS.some(keyword => lowerTitle.includes(keyword));
    if (isSensitive) {
      const creatorInfo = await axios.get(
        `${process.env.PORTAL_URL}/api/2.0/people/${createBy}`,
        {
          headers: { Authorization: `Bearer ${process.env.API_KEY}` }
        }
      );

      const creatorName = `${creatorInfo.data?.firstName || ""} ${creatorInfo.data?.lastName || ""}`.trim();
      const creatorEmail = creatorInfo.data?.email || "[unknown]";

      const adminMessage = `
Sensitive room "${roomTitle}" has been deleted.

Deleted by: ${creatorName} (${creatorEmail})
Room ID: ${roomId}
Date: ${timestamp}

Access log before deletion:
${logUrl}
      `.trim();

      await axios.post(
        process.env.NOTIFY_WEBHOOK_URL,
        {
          to: process.env.ADMIN_EMAIL,
          subject: "Sensitive Room Deleted",
          message: adminMessage
        },
        { headers: { "Content-Type": "application/json" } }
      );

      console.log("Admin notified about sensitive room deletion");
    }

    res.sendStatus(200);
  } catch (err) {
    console.error("Error handling room.deleted:", err.response?.data || err.message);
    res.sendStatus(500);
  }
});

const PORT = process.env.PORT || 3000;
app.listen(PORT, () => {
  console.log(`Webhook server listening on port ${PORT}`);
});

File: webhooks/python/samples/room.deleted.py
# Room Deletion - Notify Participants, Admin, and Save Access Log

from flask import Flask, request, abort
import hmac
import hashlib
import requests
import os
import json
import tempfile
from datetime import datetime

app = Flask(__name__)

SENSITIVE_ROOM_KEYWORDS = ["hr", "finance", "legal", "clients", "external"]

# Generate HMAC-SHA256 signature
def get_secret_hash(secret, payload):
    return "sha256=" + hmac.new(secret.encode(), payload, hashlib.sha256).hexdigest()

@app.route("/webhook", methods=["POST"])
def webhook():
    signature = request.headers.get("x-docspace-signature-256")
    computed_signature = get_secret_hash(
        os.environ.get("WEBHOOK_SECRET_KEY"),
        request.get_data()
    )

    if signature != computed_signature:
        app.logger.warning("Invalid signature")
        abort(401)

    event = request.json.get("event", {})
    payload = request.json.get("payload", {})

    trigger = event.get("trigger")
    if trigger != "room.deleted":
        return "", 200

    room_title = payload.get("title")
    room_id = payload.get("rootId")
    created_by = payload.get("createBy")
    lower_title = room_title.lower() if room_title else ""
    timestamp = datetime.utcnow().isoformat()

    app.logger.info(f'Room deleted: "{room_title}" ({room_id})')

    try:
        # Step 1: Get room participants and access types
        share_res = requests.get(
            f"{os.environ.get('PORTAL_URL')}/api/2.0/files/rooms/{room_id}/share",
            headers={"Authorization": f"Bearer {os.environ.get('API_KEY')}"}
        ).json()

        users = share_res.get("access", [])

        # Step 2: Notify all participants
        for user in users:
            email = user.get("sharedTo", {}).get("email")
            if not email:
                continue

            requests.post(
                os.environ.get("NOTIFY_WEBHOOK_URL"),
                json={
                    "to": email,
                    "subject": "Room deleted",
                    "message": f'The room "{room_title}" has been deleted. If this is unexpected, please contact your administrator.'
                },
                headers={"Content-Type": "application/json"}
            )

            app.logger.info(f"User notified: {email}")

        # Step 3: Save deletion access log
        access_log = [
            {
                "email": user.get("sharedTo", {}).get("email", "[no email]"),
                "accessLevel": user.get("access", "[unknown]")
            }
            for user in users
        ]

        log_data = {
            "roomId": room_id,
            "roomTitle": room_title,
            "deletedAt": timestamp,
            "accessBeforeDeletion": access_log
        }

        with tempfile.NamedTemporaryFile(delete=False, suffix=".json", mode="w") as tmpfile:
            json.dump(log_data, tmpfile, indent=2)
            temp_log_path = tmpfile.name

        with open(temp_log_path, "rb") as f:
            files_payload = {"file": f}
            upload_res = requests.post(
                f"{os.environ.get('PORTAL_URL')}/api/2.0/files/folder/{os.environ.get('DELETION_LOG_FOLDER_ID')}/upload",
                files=files_payload,
                headers={"Authorization": f"Bearer {os.environ.get('API_KEY')}"}
            ).json()

        os.unlink(temp_log_path)

        uploaded = upload_res.get("response", [{}])[0]
        log_url = f"{os.environ.get('PORTAL_URL')}/file/{uploaded.get('id')}"

        app.logger.info("Access log uploaded")

        # Step 4: If sensitive room - notify admin
        is_sensitive = any(keyword in lower_title for keyword in SENSITIVE_ROOM_KEYWORDS)
        if is_sensitive:
            creator_info = requests.get(
                f"{os.environ.get('PORTAL_URL')}/api/2.0/people/{created_by}",
                headers={"Authorization": f"Bearer {os.environ.get('API_KEY')}"}
            ).json()

            creator_name = f"{creator_info.get('firstName', '')} {creator_info.get('lastName', '')}".strip()
            creator_email = creator_info.get("email", "[unknown]")

            admin_message = (
                f"Sensitive room \"{room_title}\" has been deleted.\n\n"
                f"Deleted by: {creator_name} ({creator_email})\n"
                f"Room ID: {room_id}\n"
                f"Date: {timestamp}\n\n"
                f"Access log before deletion:\n{log_url}"
            )

            requests.post(
                os.environ.get("NOTIFY_WEBHOOK_URL"),
                json={
                    "to": os.environ.get("ADMIN_EMAIL"),
                    "subject": "Sensitive Room Deleted",
                    "message": admin_message
                },
                headers={"Content-Type": "application/json"}
            )

            app.logger.info("Admin notified about sensitive room deletion")

        return "", 200

    except Exception as err:
        app.logger.error(f"Error handling room.deleted: {err}")
        return "", 500

if __name__ == "__main__":
    port = int(os.environ.get("PORT", 3000))
    app.run(host="0.0.0.0", port=port)
